### ISSUE TITLE
Reading note of 'Visualizing the Loss Landscape of Neural Nets' paper by yuyi

### ISSUE THREAD

**Issue by yuyicg** [sql-machine-learning/elasticdl#181]
### empirical conclusion

- Loss landscapes for all the networks considered seem to be partitioned into a well-defined region of low loss value and convex contours, surrounded by a well-defined region of high loss value and non-convex contours.

- When networks become sufficiently deep, neural loss landscapes quickly transition from being nearly convex to being highly chaotic and skip connections promote flat minimizers and prevent the transition to chaotic behavior.

- Increased network width resulted in flat minima and wide regions of apparent convexity where sharpness of minima correlates extremely well with test error.

- Residual connections prevent the transition to chaotic behavior as depth increases.

- It is widely thought that small-batch SGD produces “flat” minimizers that generalize well, while large
batches produce “sharp” minima with poor generalization.

- The “random directions” plotting approach fails to capture the intrinsic geometry of loss surfaces because of the `scale invariance(e.g. ReLU)` in network weights, and cannot be used to compare the geometry of two different minimizers or two different networks, so the filter-wise normalization was proposed.

- To capture variation in trajectory of optimizer overcoming high probability of orthogonality in high dimensional space, author proposed non-random (and carefully chosen) directions (PCA based approach).

### proposed visualization algorithm

- Plotting the loss function using filter-wise normalization:
plot the funtion of `f (α, β) = L(θ ∗ + αδ + βη)` for 2D surface.
δ, η are random Gaussian direction vector with dimensions compatible with θ, and are normalized by formula:
![screenshot from 2019-01-21 15-03-29](https://user-images.githubusercontent.com/5202350/51457840-e587fe80-1d8d-11e9-8b9d-debf90f1edc5.png)
where d i,j represents the jth filter (not the jth weight) of the ith layer of d, and ||*|| denotes the Frobenius norm.

- Plotting trajectory of optimizer using PCA directions which apply PCA to the matrix `M = [θ 0 −θ n ; · · · ; θ n−1 −θ n ]` where θ i denote model parameters at epoch i, and then select the two most explanatory directions.

### findings

- This paper studied the plotting method of the geometry of single local minima and the relationship between sharpness and generalization ability, but does not compare multiple local minima and how to find the global optimality.

- Our next work may need to explore the possible differences of loss value between different local optima and the possibility of finding global optima(could use the proposed plotting method), because for now it seems that the loss values of different local optima are very close in the state-of-art network models.

- corresponding plotting codes used in paper [https://github.com/tomgoldstein/loss-landscape](https://github.com/tomgoldstein/loss-landscape)
